encrypt/ipsec_rotate_key.go

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package encrypt
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/cilium/cilium-cli/defaults"
+)
+
+type ipsecKey struct {
+	spi        int
+	spiSuffix  bool
+	algo       string // Purposefully ambiguous here because of modes like GCM
+	key        string
+	size       int
+	cipherMode string
+	cipherKey  string
+}
+
+// IPsecRotateKey rotates IPsec key.
+func (s *Encrypt) IPsecRotateKey(ctx context.Context) error {
+	ctx, cancelFn := context.WithTimeout(ctx, s.params.WaitDuration)
+	defer cancelFn()
+
+	secret, err := s.client.GetSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch IPsec secret: %s", err)
+	}
+
+	key, err := ipsecKeyFromString(string(secret.Data["keys"]))
+	if err != nil {
+		return err
+	}
+
+	newKey, err := key.rotate()
+	if err != nil {
+		return fmt.Errorf("failed to rotate IPsec key: %s", err)
+	}
+
+	patch := []byte(`{"stringData":{"keys":"` + newKey.String() + `"}}`)
+	_, err = s.client.PatchSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to patch IPsec secret with new key: %s", err)
+	}
+
+	_, err = fmt.Printf("IPsec key successfully rotated, new key SPI: %d\n", newKey.spi)
+	return err
+}
+
+func (k ipsecKey) String() string {
+	spiSuffix := ""
+	if k.spiSuffix {
+		spiSuffix = "+"
+	}
+	if k.cipherMode == "" {
+		return fmt.Sprintf("%d%s %s %s %d", k.spi, spiSuffix, k.algo, k.key, k.size)
+	}
+	return fmt.Sprintf("%d%s %s %s %s %s", k.spi, spiSuffix, k.algo, k.key, k.cipherMode, k.cipherKey)
+}
+
+var (
+	ipsecKeyRegex    = regexp.MustCompile(`^([[:digit:]]+\+?)[[:space:]](\S+)[[:space:]]([[:alnum:]]+)[[:space:]]([[:digit:]]+)$`)
```

Comment:
Many security issues (CVEs) are the result of imprecise parsing, with [Regex often being implicated](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=regex).

The related thread model here is that an attacker can somehow control the ipsec key Kubernetes secret, which may make such hardening out of scope, although it is something that should be considered.

Reply:
So, do you think we should avoid using `regexp` and parse it like a CSV string instead?

Reply:
My gut feeling is that this is more precise than a simple field split, and is fine for now given the threat model.